src/Raven.Server/Utils/Metrics/MeterMetric.cs

```diff
@@ -131,42 +104,150 @@ public void Mark()
 
         public DynamicJsonValue CreateMeterData(bool allResults = false, bool filterEmpty = true)
         {
-            var meterValue = this;
-
-            var r = new DynamicJsonValue
+            var result = new DynamicJsonValue
             {
-                ["Current"] = Math.Round(meterValue.OneSecondRate, 1),
-                ["Count"] = meterValue.Count,
-                ["MeanRate"] = Math.Round(meterValue.MeanRate, 1),
-                ["OneMinuteRate"] = Math.Round(meterValue.OneMinuteRate, 1),
-                ["FiveMinuteRate"] = Math.Round(meterValue.FiveMinuteRate, 1),
-                ["FifteenMinuteRate"] = Math.Round(meterValue.FifteenMinuteRate, 1)
+                ["Current"] = Math.Round(OneSecondRate, 1),
+                ["Count"] = Count,
+                ["MeanRate"] = Math.Round(MeanRate, 1),
+                ["OneMinuteRate"] = Math.Round(OneMinuteRate, 1),
+                ["FiveMinuteRate"] = Math.Round(FiveMinuteRate, 1),
+                ["FifteenMinuteRate"] = Math.Round(FifteenMinuteRate, 1)
             };
 
             if (allResults == false)
-                return r;
-
-            var results = new DynamicJsonValue();
-            r["Raw"] = results;
-            var index = Volatile.Read(ref _index) % _m15Rate.Length;
-            var current = DateTime.UtcNow;
-            var now = new TimeSpan(current.Hour, current.Minute, current.Second);
-            var oneSec = TimeSpan.FromSeconds(1);
-            for (int i = index; i >= 0; i--)
+                return result;
+
+            var nowNs = _now();
+            var nowUtc = DateTime.UtcNow;
+            var bucketSeconds = BucketDurationNanoseconds / (double)Clock.NanosecondsInSecond;
+
+            var buckets = new List<(long Start, double Rate)>();
+            for (int i = 0; i < BucketCount; i++)
             {
-                var d = _m15Rate[i];
-                if (filterEmpty == false || Math.Abs(d) > double.Epsilon)
-                    results[now.ToString()] = Math.Round(d, 1);
-                now = now - oneSec;
+                ref var bucket = ref _buckets[i];
```

Comment:
Why by-ref here? If you copied _buckets[i] to local, you'd have it's snapshot and no further volatiles would be needed?